### PR TITLE
feat(api-version): use api version in request strategies

### PR DESCRIPTION
**Pull request by David-Henner**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

In this PR we update the request creation logic in order to use the current api version (`APIVersion.current`) instead of using `v0` by default. Since `APIVersion.current` returns an optional `APIVersion`, we need to handle the `nil` case.

Some of the functions used to create requests receive a completion handler. In these cases, if the version is `nil` we pass a `failure` result into the completion handler. Failures (mostly) need an `Error` as associated value. When defined, we associate the `unknown` error. Otherwise, we create a special case for the `missingAPIVersion` error. 

The right strategy to handle the nullability of `APIVersion.current` isn't entirely clear yet. As well as the necessity to handle nullability, since we expect the API version to always be fetched when the app enters the foreground.

Another change concerns the `FederationAware` request strategies. There we cannot simply keep the existing behaviour. So we switch over the API version. For `v0` we keep creating requests for legacy endpoints. For `v1`, since federation endpoints are supported regardless of whether or not federation is enabled, we create requests for federation endpoints. However, when federation is disabled, we use the domain from `APIVersion`. 

Now, in practice, we expect the domain from the subject of the request to be `nil` when federation is disabled. So, we don't directly check the federation flag. Instead, if the domain is `nil` we fall back onto the `APIVersion` domain. 

But, if the `APIVersion` domain is `nil`, it is also unclear which strategy we should follow. Since we expect the domain to be fetched the same way as the API version (mentioned earlier). Should we fallback onto the legacy endpoints? Or just return a `nil` request? 

If we fallback, should it only be when federation is disabled? Or in both cases?

### Testing

Updated tests to:
- Check that the path is set correctly for different API versions
- Set api version when needed

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: Source/Synchronization/Strategies/SearchUserImageStrategy.swift [hunk 1]
```diff
@@ -132,18 +130,22 @@ public class SearchUserImageStrategy: AbstractRequestStrategy, FederationAware {
...
     }
 
-    func request(for assetKeys: SearchUserAssetKeys, size: ProfileImageSize, user: UUID) -> ZMTransportRequest? {
+    func request(for assetKeys: SearchUserAssetKeys, size: ProfileImageSize, user: UUID, apiVersion: APIVersion) -> ZMTransportRequest? {
         if let key = size == .preview ? assetKeys.preview : assetKeys.complete {
             let path: String
 
-            if useFederationEndpoint,
-               requestedUserDomain.keys.contains(user),
-               let domain = requestedUserDomain[user] {
```
[the remark this answers is not in the document]
Comment (David-Henner) [reply Source/Synchronization/Strategies/SearchUserImageStrategy.swift:141]: Couldn't the fallback be using the domain from `APIVersion` like I did below ?

File: Source/Synchronization/Strategies/UserImageAssetUpdateStrategy.swift [hunk 2]
```diff
@@ -171,12 +169,19 @@ public final class UserImageAssetUpdateStrategy: AbstractRequestStrategy, ZMCont
             remoteId = user.completeProfileAssetIdentifier
         }
         guard let assetId = remoteId else { return nil }
+
         let path: String
-        if useFederationEndpoint, let domain = user.domain {
```
[the remark this answers is not in the document]
Comment (David-Henner) [reply Source/Synchronization/Strategies/UserImageAssetUpdateStrategy.swift:175]: same comment regarding falling back on the domain from `APIVersion`

File: Source/Data Model/Conversation+Deletion.swift [hunk 3]
```diff
@@ -77,11 +77,15 @@ extension ZMConversation {
 struct ConversationDeletionRequestFactory {
 
     static func requestForDeletingTeamConversation(_ conversation: ZMConversation) -> ZMTransportRequest? {
-        guard let conversationId = conversation.remoteIdentifier, let teamRemoteIdentifier = conversation.teamRemoteIdentifier else { return nil }
+        guard
```
[the remark this answers is not in the document]
Comment (David-Henner) [reply Source/Data Model/Conversation+Deletion.swift:80]: I don't think so, it's called through an extension of `ZMConversation` and then the request is enqueued on the transport session

File: Source/Synchronization/Strategies/SearchUserImageStrategy.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (johnxnguyen) [reply Source/Synchronization/Strategies/SearchUserImageStrategy.swift:141]: Ah right. Yeah I guess that could work.

**Review by johnxnguyen — APPROVED** [review]

**Review by agisilaos — APPROVED** [review]

**Comment by agisilaos** [comment]
We have a a lot of test failures. Are we going to take care of those in a separate PR?

**Pull request merged**